pygmt/projection.py

```diff
@@ -0,0 +1,590 @@
+#!/usr/bin/env python
+
+"""
+Contains the projections supported by GMT, and the necessary mechanisms
+to create a projection and output a valid GMT projection string.
+
+>>> from pygmt import projection
+>>> proj = projection.LambertAzimuthalEqualArea(central_longitude=30, central_latitude=-20, horizon=60, width=8, unit="i")
+>>> proj
+LambertAzimuthalEqualArea(central_longitude=30, central_latitude=-20, horizon=60, width=8, unit='i', center=[30, -20])
+>>> print(proj)
+A30/-20/60/8i
+"""
+
+from enum import Enum
+import attr
+
+
+class Supported(Enum):
+
+    """
+    The supported projections and their GMT code.
+    """
+
+    UNDEFINED = ""
+    LAMBERT_AZIMUTH_EQUAL_AREA = "A"  # DONE
+    ALBERS_CONIC_EQUAL_AREA = "B"  # DONE
+    CASSINI_CYLINDRICAL = "C"  # DONE
+    CYLINDRICAL_STEROGRAPHIC = "JCyl_stere/"  # includes `/` according to https://docs.generic-mapping-tools.org/latest/proj_codes.html  # DONE
+    EQUIDISTANT_CONIC = "JD"  # DONE
+    AZIMUTHAL_EQUIDISTANT = "E"  # DONE
+    AZIMUTHAL_GNOMIC = "F"  # DONE
+    AZIMUTHAL_ORTHOGRAPHIC = "G"  # DONE
+    GENERAL_PERSPECTIVE = "G"  # DONE
+    HAMMER_EQUAL_AREA = "H"
+    SINUSOIDAL_EQUAL_AREA = "I"
+    MILLER_CYLINDRICAL = "J"
+    ECKERT_IV_EQUAL_AREA = "Kf"
+    ECKERT_VI_EQUAL_AREA = "Ks"
+    LAMBERT_CONIC_CONFORMAL = "L"
+    MERCATOR_CYLINDRICAL = "M"  # DONE
+    ROBINSON = "N"
+    OBLIQUE_MERCATOR_1 = "Oa"
+    OBLIQUE_MERCATOR_2 = "Ob"
+    OBLIQUE_MERCATOR_3 = "Oc"
+    POLAR = "P"
+    POLYCONIC = "Poly"
+    EQUIDISTANT_CYLINDRICAL = "Q"
+    WINKEL_TRIPEL = "R"
+    GENERAL_STEREOGRAPHIC = "S"  # DONE
+    TRANSVERSE_MERCATOR = "T"
+    UNIVERSAL_TRANSVERSE_MERCATOR = "U"
+    VAN_DER_GRINTEN = "V"
+    MOLLWEIDE = "W"
+    LINEAR = "X"
+    CYLINDRICAL_EQUAL_AREA = "Y"  # DONE
+
+
+@attr.s()
+class _Projection:
+
+    """
+    Base class for all projections.
+    """
+
+    # private; we don't want the user to care or know about
+    _fmt: str = attr.ib(init=False, repr=False, default="{_code}")
+    _code: str = attr.ib(init=False, repr=False, default=Supported.UNDEFINED.value)
```

Comment:
`_code` and `_fmt` are the main things that child classes need to specify, right? If so, they should abstract in the base class. To do that, the `_Projection` and `_Azimuthal` et al should probably be [abstract base classes](https://docs.python.org/3/library/abc.html).

Reply:
I'm not sure how well an `ABC` will work with a child class being defined via `attrs`.  From the docs, `attrs` trawls the class hierarchy and collects all attributes, and then writes its' own methods to access the required attributes.

So setting an attribute on a child class on a property defined via an `ABC` fails.

eg:
```python
from abc import ABCMeta, abstractproperty
@attr.s
class _Proj:
    __metaclass__ = ABCMeta

    @abstractproperty
    def longitude(self):
        pass

    @abstractproperty
    def _code(self):
        pass

@attr.s
class MyProj(_Proj):
    longitude: float = attr.ib()
    _code: str = attr.ib(default="X", init=False, repr=False)
```

Calling:
```python
p = MyProj(145)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<attrs generated init __main__.MyProj-2>", line 3, in __init__
AttributeError: can't set attribute
```

A way around this would be to create a different attribute and define a property that returns it, but I think it then adds additional complexity to the class definition when you need to define the attribute.

eg
```python
@attr.s
class MyProj(_Proj):
    _longitude = attr.ib()
    __code = attr.ib(default="X", init=False, repr=False)

    @property
    def longitude(self):
        return self._longitude

    @property
    def _code(self):
        return self.__code
```

Was this along the lines of what you meant when defining the base classes to be an [abstract base class](https://docs.python.org/3/library/abc.html)?
I'll admit I haven't done much with them since making the move to [attrs](https://www.attrs.org/en/stable/index.html).